### PR TITLE
Policy: fix ToServices performance regression

### LINKED ISSUE #35273: Inefficient resolution of toServices in CNPs
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

When performing CNP-scalability testing with "regular" policies that do not have `toServices` rules I've noticed that a significant amount of memory allocations come from [resolveServices](https://github.com/cilium/cilium/blame/d73443647bd44bca9ea883d68a2874db813e93f0/pkg/policy/k8s/service.go#L117-L146)
pprof with memory allocations:
Image

`resolveToServices` is called for each CNP regardless if the CNP has the `toServices` rule or not. On top of that, we iterate through all services and rules in CNP: https://github.com/cilium/cilium/blame/d73443647bd44bca9ea883d68a2874db813e93f0/pkg/policy/k8s/service.go#L117-L146

This is super inefficient and wasteful. Instead, we should:
cache information if a particular CNP affects services and only process it when:
- CNP has `toServices` rule
- CNP previously had `toServices` rule, but the rule was removed

### How can we reproduce the issue?

N/A

### Cilium Version

~main branch

### Kernel Version

N/A

### Kubernetes Version

N/A

### Regression

Potentially regression from https://github.com/cilium/cilium/pull/31062 but I haven't checked code in detail how it was handled previously.

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by squeed**
This PR contains two commits:

1: Make service iteration more efficient. It turns out that calling `eps.Endpoints()` is very allocation-heavy. Before this change, it was called for every service when iterating, which was unnecessary. Instead, leave this to the caller, who can do so only when necessary

2: Skip policy ToServices translation when policy has no ToServices selectors. We already track whether or not policies have a ToServices selector. When one does not, we do not need to try and translate services.

Fixes: #35273

```release-note
Fixes a performance regression when ingesting network policies in clusters with large numbers of Services.
```

### CODE REVIEW

File: pkg/policy/k8s/cilium_network_policy.go [hunk 1]
```diff
@@ -134,7 +134,9 @@ func (p *policyWatcher) resolveCiliumNetworkPolicyRefs(
 	translatedCNP := cnp.DeepCopy()
 
 	// Resolve ToService references
-	p.resolveToServices(key, translatedCNP)
+	if _, exists := p.toServicesPolicies[key]; exists {
+		p.resolveToServices(key, translatedCNP)
```
Comment (marseel) [inline pkg/policy/k8s/cilium_network_policy.go:145]: I am happy about it from the scalability point of view, but I am not sure if it's correct. 
`toServicesPolicies` is updated in the parent function here: https://github.com/cilium/cilium/blob/9b7da39de54453b229a88c5ff5dbcf09037f1d5f/pkg/policy/k8s/cilium_network_policy.go#L77-L81

Imagine that someone removes `toServices` in CNP. Shouldn't we still perform `p.clearCNPForService(key, svcID)` for whatever reason it is there?

Comment (squeed) [reply pkg/policy/k8s/cilium_network_policy.go:145]: very good catch!

Comment (squeed) [reply pkg/policy/k8s/cilium_network_policy.go:145]: OK, that should do it.

**Review by marseel — APPROVED** [review]
I have no power here, but looks great to me :) 

Thanks!

**Review by doniacld — APPROVED** [review]
lgtm

**Review by tommyp1ckles — APPROVED** [review]

**Pull request merged**